Replace tf config names in error strings

VenelinMartinov opened the pull request:
Should address https://github.com/pulumi/pulumi-terraform-bridge/issues/1424

The error referred to the tf config names:
```
must provide exactly one of "api_key", "api_token" or "api_user_service_key".
```

And it's now this:
```
must provide exactly one of "cloudflare:apiKey", "cloudflare:apiToken" or "cloudflare:apiUserServiceKey".
```

I still need to do the same for pf. 
Happy to move the code around if there's a better place for it but it looks like `check_failures.go` doesn't do what we need here. The errors are plain strings raised from the terraform provider and don't follow the usual validation logic in tf - looks like the upstream provider is missing a validation line to mark the three configs as `ExactlyOneOf`. This means that the validation doesn't happen in `CheckConfig` but instead happens during `Configure`.

I'm happy to also raise a PR upstream if we think that'd help.

t0yv0 commented on pkg/tfbridge/provider.go:
```diff
@@ -517,7 +517,7 @@ func (p *Provider) Configure(ctx context.Context,
 	configMap, err := configEnc.UnmarshalProperties(req.GetArgs())
 	if err != nil {
-		return nil, err
+		return nil, ReplaceErrorProperties(err, p.info)
```
I don't believe that configEnc.UnmarshalProperties will ever surface missing config errors; is this needed?

t0yv0 commented on pkg/tfbridge/replaceErrorProperties.go:
```diff
@@ -0,0 +1,48 @@
+
+	renamedConfigProperties := renames["renamedConfigProperties"]
+	var configPropertyRenames map[string]string = make(map[string]string)
+	for k, v := range renamedConfigProperties.(map[string]interface{}) {
```
We'll likely need to remove rename tables soon from the provider. They're equivalent in functionality to TerraformToPulumiNameV2 with some caveats about nested properties. I recommend you reuse NewCheckFailurePath as that should cover this use case a well as nested ones.

iwahbe replied:
Also, we don't require all providers to have rename tables in use.

VenelinMartinov replied:
I'm not sure that `NewCheckFailurePath` works here, as I am searching the terraform error string for the terraform names, and only then converting it. This strategy needs a list of terraform names to search for, which seems to only be available in the rename table.

The solution here might be to raise a PR upstream to add the validation and use `check_failures.go` after.

t0yv0 replied:
Well, what would it take to convince you? The list of terraform provider configuration property names is indeed available without the rename tables if needed; it can be obtained from the Provider shim.

iwahbe commented on pkg/tfbridge/replaceErrorProperties.go:
```diff
@@ -0,0 +1,48 @@
+	return s
+}
+
+func getConfigReplacements(info ProviderInfo) (map[string]string, error) {
```
It would be great if we could build this from `(map[string]*SchemaInfo, shim.SchemaMap)`. That would allow us to have the same error correction function run for the provider itself and for each resource.

iwahbe commented on pkg/tfbridge/replaceErrorProperties.go:
```diff
@@ -0,0 +1,48 @@
+	return strings.ReplaceAll(s, old, new)
+}
+
+func ReplaceSubstrings(s string, replacements map[string]string) string {
```
I think all the functions in this module (except the entry point) should be private.

VenelinMartinov commented:
>Can you shed some more light on why it's not happening in CheckConfig but only in Configure? Is that certain? Are you saying thet p.tf.Validate(config) does get called but accepts this config without erroring out?

Yeah, `CheckConfig` accepts the config without erroring out. The check that at least one of the configs is set is run during `Configure` [link](https://github.com/cloudflare/terraform-provider-cloudflare/blob/54518a44cc0b7fba844b802e95e253893c3f65b5/internal/framework/provider/provider.go#L305). The [terraform schema](https://github.com/cloudflare/terraform-provider-cloudflare/blob/54518a44cc0b7fba844b802e95e253893c3f65b5/internal/framework/provider/provider.go#L92C3-L92C3) does not have the validation logic about these keys.

The TF validation logic can have `stringvalidator.ExactlyOneOf(path.MatchRoot(consts.APIKeySchemaKey), path.MatchRoot(consts.APITokenSchemaKey), path.MatchRoot(consts.APIUserServiceKeySchemaKey))`, which is then correctly passed down to pulumi and we do indeed error during `CheckConfig` if that line is added upstream.

t0yv0 commented:
Thank you very much for providing the references, this sheds a lot of light on the problem. It appears that this provider bypasses standard TF validation facilities and does ad-hoc validation in Configure.

Fixing it in the provider instead of the bridge is not unreasonable IMO because it might be a rare error. It could be a PreCheckCallback kind of thing that does pulumi-level validation. 

On the other hand we don't know how often this happens but it seems like it could happen a fair deal.

 In that case the suggested fix here of scanning errors and rewriting them also makes sense. 

If you decide to go ahead, I'm happy to help you land this.

First please address the PR feedback to not rely on rename tables
 
https://github.com/pulumi/pulumi-terraform-bridge/blob/master/pkg/tfshim/shim.go#L165 has Provider.

provider.Schema().Range(...) will give you top-level Terraform property names for the provider configuration values. 

If you don't want to tackle properties deeper than the top-level (nested properties) in this PR that's fine with me as they are much more rare. Just please document in a comment linking a ticket for later.

NewCheckFailurePath can translate nested or top-level TF property names to Pulumi names. If you're only interested in top-level names, this is equivalent to calling TerraformToPulumiNameV2 which is currently the definitive API for name translation.

VenelinMartinov commented:
Thanks a lot for the help! I believe I addressed most of the comments.

I've raised https://github.com/pulumi/pulumi-terraform-bridge/issues/1533 for any further work. What's a good example of such a nested property?

I'll also add some tests here.

VenelinMartinov commented:
Cloudflare has a config property called simply `email`. Are we concerned about being overzealous here? AWS has configs called `region` and `profile`.

iwahbe commented:
> Cloudflare has a config property called simply `email`. Are we concerned about being overzealous here? AWS has configs called `region` and `profile`.
> 
> I'm wondering if these might not yield more confusing errors.
> 
> Should we have an opt-out list of generic sounding names?

I thought that this PR was scoped to quoted names. `region` might refer to many things, but `"region"` is almost definitely a provider field.

t0yv0 commented:
Wouldn't "region" and "profile" translate to themselves ("region" and "profile")?

VenelinMartinov commented:
> I thought that this PR was scoped to quoted names. region might refer to many things, but "region" is almost definitely a provider field.
Great point @iwahbe! I think I meant to limit it to quoted fields but didn't implement it like that.

@t0yv0 it'd get translated to `aws:region` and `aws:profile` since we add the provider prefix so that the config prints correctly.

t0yv0 commented on pkg/tfbridge/replaceErrorProperties.go:
```diff
@@ -0,0 +1,40 @@
+	schema shim.SchemaMap,
+) map[string]string {
+	renames := make(map[string]string)
+	schema.Range(func(key string, value shim.Schema) bool {
```
The for-loop is not a great strategy.. If you have foo and fooProp it might replace twice. Why don't you build a regex and replace in one pass.

t0yv0 commented:
> it'd get translated to aws:region and aws:profile since we add the provider prefix so that the config prints correctly.

How about we drop the 'aws:' prefix? It's not useful anyway in the case the error is coming from an Explicit Provider.

t0yv0 commented on pkg/tfbridge/replaceErrorProperties.go:
```diff
@@ -0,0 +1,40 @@
+
+// ReplaceConfigProperties replaces all Terraform config property names
+// in the given message with their Pulumi equivalents.
+// This only works for top-level properties currently.
```
Unfortunate to need another top-level public export here. Please make a comment here this is not intended for consumption by provider authors as we're only exposing this to reuse for Plugin framework bridge.

t0yv0 commented:
Looking good. Can we add tests for Configure to demonstrate this is working? TYVM.

t0yv0 commented on pkg/tfbridge/provider_test.go:
```diff
@@ -1425,6 +1423,37 @@ func TestConfigure(t *testing.T) {
+			{
+			  "method": "/pulumirpc.ResourceProvider/Configure",
+			  "request": {"acceptResources": true},
+			  "errors": "1 error occurred:\n\t* some error with \"configProperty\" and \"CONFIG!\" but not config\n\n"
```
👍

t0yv0 commented on pf/tfbridge/provider_diagnostics.go:
```diff
@@ -37,7 +38,16 @@ func (p *provider) processDiagnostics(diagnostics []*tfprotov6.Diagnostic) error
 				prefix = fmt.Sprintf("[%s] ", d.Attribute.String())
 			}
-			return fmt.Errorf("%s%s: %s", prefix, d.Summary, d.Detail)
+			summary, summaryErr := tfbridge.ReplaceConfigProperties(d.Summary, p.info.Config, p.info.P.Schema())
```
I'm curious why this is applied at procesDiagnostics, which is applying to *everything*, Configure, Create, you name it. This is a very wide net to cast. That may be the right thing to do but I'm curious is this intentional? For SDKv2 based providers the change seems a lot more surgical, only affecting Configure.

VenelinMartinov replied:
Yeah, good point. Thanks for pointing that out - it was not intentional, should now be corrected.

Could be the right thing to do for other methods but it might be better to make the change once we have a concrete reason.

t0yv0 commented:
The tests look good. This is looking like it can ship to me, one last nit is the question why SDKv2 is more surgical than PF.
